### PR TITLE
chore!: enhance TX error handling and message formatting

### PR DESCRIPTION
**Pull request by Torres-ssf**
- Close: https://github.com/FuelLabs/fuels-ts/issues/823

## Breaking Changes:

- The `ScriptResultDecoderError` along with all revert error-related classes and its helpers [were removed](https://github.com/FuelLabs/fuels-ts/pull/1895/commits/1f735a50824bc81b8da388e5c1807214dd85ee08).

### CODE REVIEW

**Comment by Torres-ssf** [comment]
Blocked by: https://github.com/FuelLabs/fuels-ts/pull/1940

**Comment by fuel-service-user** [comment]
✨ A PR has been created under the `rc-1895` tag on `fuels-wallet` repo.
https://github.com/FuelLabs/fuels-wallet/pull/1204

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  logs: Array<unknown>;
+}
+
+export const extractTxError = (params: IExtractTxError): FuelError => {
+  const { receipts, status, logs } = params;
+
+  const wasPanic = receipts.some(({ type }) => type === ReceiptType.Panic);
+
+  let err =
+    status?.type === 'FailureStatus' && wasPanic
```
Comment (nedsalk) [inline packages/account/src/providers/utils/extract-tx-error.ts:93]: The `status?.type === 'FailureStatus'` check seems redundant because we're inside of `extractTxError` which implies that we're dealing with a failed transaction.

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      ? assemblePanicError(status)
+      : assembleRevertError(receipts, logs);
```
Comment (nedsalk) [inline packages/account/src/providers/utils/extract-tx-error.ts:95]:
```suggestion
  const isPanic = receipts.some(({ type }) => type === ReceiptType.Panic);

  let err =
    status?.type === 'FailureStatus' && isPanic
      ? assemblePanicError(status)
      : assembleRevertError(receipts, logs);
```

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      case FAILED_ASSERT_NE_SIGNAL: {
+        const sufix =
+          logs.length >= 2 ? ` comparing ${stringify(logs[1])} and ${stringify(logs[0])}.` : '.';
+
+        errorMessage = `The transaction reverted because of an "assert_ne" statement${sufix}`;
+        break;
+      }
+
+      case FAILED_ASSERT_SIGNAL:
+        errorMessage = `The transaction reverted because of an "assert" statement failed to evaluate to true.`;
```
Comment (nedsalk) [inline packages/account/src/providers/utils/extract-tx-error.ts:66]:
```suggestion
        errorMessage = `The transaction reverted because an "assert" statement failed to evaluate to true.`;
```
This'll break some tests if merged directly as a suggestion.

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        break;
+
+      case FAILED_TRANSFER_TO_ADDRESS_SIGNAL:
+        errorMessage = `The transaction reverted because missing "OutputChange"(s).`;
```
Comment (nedsalk) [inline packages/account/src/providers/utils/extract-tx-error.ts:70]:
```suggestion
        errorMessage = `The transaction reverted because it's missing an "OutputChange"(s).`;
```

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  const revertReceipt = receipts.find(({ type }) => type === ReceiptType.Revert) as ReceiptRevert;
+
+  if (revertReceipt) {
+    const reasonHex = bn(revertReceipt.val).toHex();
+
+    switch (reasonHex) {
+      case FAILED_REQUIRE_SIGNAL: {
+        errorMessage = `The transaction reverted because of a "require" statement has thrown ${
+          logs.length ? stringify(logs[0]) : 'an error.'
+        }.`;
```
Comment (nedsalk) [inline packages/account/src/providers/utils/extract-tx-error.ts:45]:
```suggestion
        errorMessage = `The transaction reverted because a "require" statement has thrown ${
          logs.length ? stringify(logs[0]) : 'an error.'
        }.`;
```

File: packages/account/src/providers/transaction-response/transaction-response.ts [hunk 6]
```diff
@@ -274,10 +285,17 @@ export class TransactionResponse {
...
+        logs = [],
+        gqlTransaction: { status },
+      } = result;
+
+      throw extractTxError({
+        receipts,
+        logs,
+        status,
+      });
     }
```
Comment (nedsalk) [inline packages/account/src/providers/transaction-response/transaction-response.ts:300]:
```suggestion
```

This code block is now redundant because it's handled in `this.waitForResult`.

File: packages/account/src/providers/transaction-response/transaction-response.ts [hunk 7]
```diff
@@ -260,6 +258,19 @@ export class TransactionResponse {
       transactionResult.logs = logs;
     }
 
+    if (transactionResult.isStatusFailure) {
```
Comment (nedsalk) [inline packages/account/src/providers/transaction-response/transaction-response.ts:261]: _Commenting for reference_

This logic should be moved into `this.waitForStatusChange` when implementing #1644.

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  return errorMessage;
+};
+
+interface IExtractTxError {
+  receipts: Array<TransactionResultReceipt>;
+  status?: GqlTransactionStatusFragmentFragment | null;
+  logs: Array<unknown>;
+}
+
+export const extractTxError = (params: IExtractTxError): FuelError => {
```
Comment (petertonysmith94) [inline packages/account/src/providers/utils/extract-tx-error.ts:87]:
```suggestion
/** @hidden */
export const extractTxError = (params: IExtractTxError): FuelError => {
```

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  FAILED_TRANSFER_TO_ADDRESS_SIGNAL,
+  PANIC_REASONS,
+  PANIC_DOC_URL,
+} from '@fuel-ts/transactions/configs';
+
+import type { GqlTransactionStatusFragmentFragment } from '../__generated__/operations';
+import type { TransactionResultReceipt } from '../transaction-response';
+import type { FailureStatus } from '../transaction-summary';
+
+export const assemblePanicError = (status: FailureStatus) => {
```
Comment (petertonysmith94) [inline packages/account/src/providers/utils/extract-tx-error.ts:19]:
```suggestion
/** @hidden */
export const assemblePanicError = (status: FailureStatus) => {
```

File: packages/account/src/providers/utils/extract-tx-error.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+  if (PANIC_REASONS.includes(status.reason)) {
+    errorMessage = `${errorMessage}\n\nYou can read more about this error at:\n\n${PANIC_DOC_URL}#variant.${status.reason}`;
+  }
+
+  return errorMessage;
+};
+const stringify = (obj: unknown) => JSON.stringify(obj, null, 2);
+
+export const assembleRevertError = (
```
Comment (petertonysmith94) [inline packages/account/src/providers/utils/extract-tx-error.ts:30]:
```suggestion
/** @hidden */
export const assembleRevertError = (
```

File: packages/account/src/providers/utils/extract-tx-error.ts (hunk 1 above)
Comment (Torres-ssf) [reply packages/account/src/providers/utils/extract-tx-error.ts]: Looks a little bit redundant to me as well.

However the `status` will not always be present when calling this function, since it is also used to decode errors from dry-runs as well, where only receipts are present ( this changes on fuel-core `0.23.0` ). 

The if statement is to satisfy TS, since `reason` is always present within a `FailureStatus`

**Review by arboleya — DISMISSED** [review]

**Review by nedsalk — APPROVED** [review]

**Review by petertonysmith94 — DISMISSED** [review]

**Review by danielbate — APPROVED** [review]

**Review by petertonysmith94 — APPROVED** [review]
LGTM 🚀

File: packages/transactions/src/configs.ts [hunk 11]
```diff
@@ -35,13 +35,64 @@ export const FAILED_REQUIRE_SIGNAL = '0xffffffffffff0000';
 // Revert with this value for a failing call to `std::asset::transfer_to_address`.
 export const FAILED_TRANSFER_TO_ADDRESS_SIGNAL = '0xffffffffffff0001';
 
-// Revert with this value for a failing call to `std::message::send_message`.
-export const FAILED_SEND_MESSAGE_SIGNAL = '0xffffffffffff0002';
```
Comment (Dhaiwat10) [inline packages/transactions/src/configs.ts:39]: Why was this removed?

**Review by Dhaiwat10 — APPROVED** [review]

File: packages/transactions/src/configs.ts (hunk 11 above)
Comment (Torres-ssf) [reply packages/transactions/src/configs.ts:39]: @Dhaiwat10 This revert type was removed from Sway
